### PR TITLE
modifications to allow 'Return-Path' to be specified to be used for r…

### PR DESCRIPTION
**Pull request by mgage**
…eporting bounce back messages
     $mail{set_return_path} is set in the SMTP configuration section of site.conf
	my $return_path_for_errors = $ce->{mail}->{set_return_path}; 
	# $set_return_path is the address used to report returned email. It is an argument 
	# used in sendmail() (aka Email::Simple::sendmail).
	# For arcane historical reasons sendmail  actually sets the field "MAIL FROM" and the smtp server then 
	# uses that to set "Return-Path".
	# references:
	#  stackoverflow: https://stackoverflow.com/questions/1235534/what-is-the-behavior-difference-between-return-path-reply-to-and-from
	#  Email::Simple: https://metacpan.org/pod/Email::Sender::Manual::QuickStart#envelope-information
	#

### CODE REVIEW

**Comment by mgage** [comment]
Don't merge yet.  
To test :
1. Modify site.conf to contain the $mail{set_return_path} variable from site.conf.dist. 
2. send a Feedback message from your webwork course.
3. Inspect the message sent to the instructor by viewing all off the headers (I used the "raw" presentation in Apple Mail ).  You should see "Return-Path" with a value of noreply@.....

If you can determine whether this solves any issues you've had with spam filters on your smtp servers.

If all of this works there are three more places where modifications need to be made. 
after line 893 SendMail.pm
after line 949 SendMail.pm
after line 619 ProblemUtil.pm

File: lib/WeBWorK/ContentGenerator/Feedback.pm [hunk 1]
```diff
@@ -304,22 +303,6 @@ $emailableURL
...
-		# 	print CGI::p(CGI::a({-href => $returnURL}, "Return to your work"));
-		# 	print CGI::pre(wrap("", "", $feedback));
-		# } else {
-		# 	$self->feedbackForm($user, $returnURL,
-		# 		"Failed to send message ($result): $Mail::Sender::Error");
-		# }
-	} else {
-		# just print the feedback form, with no message
-		$self->feedbackForm($user, $returnURL, "");
 	}
```
Comment (drgrice1) [inline lib/WeBWorK/ContentGenerator/Feedback.pm:306]: The else clause is still needed here.  The "Email Instructor" page now opens with no contents with this.

**Review by drgrice1 — CHANGES_REQUESTED** [review]
This is good for a hotfix if there are those that need it.

There are some things that need to be fixed though.

File: lib/WeBWorK/ContentGenerator/Feedback.pm [hunk 2]
```diff
@@ -217,28 +217,27 @@ sub body {
...
+		my $return_path_for_errors = $ce->{mail}->{set_return_path}; 
+	# createEmailSenderTransportSMTP is defined in ContentGenerator
+	# $set_return_path is the address used to report returned email. It is an argument 
+	# used in sendmail() (aka Email::Simple::sendmail).
+	# For arcane historical reasons sendmail  actually sets the field "MAIL FROM" and the smtp server then 
+	# uses that to set "Return-Path".
+	# references:
+	#  stackoverflow: https://stackoverflow.com/questions/1235534/what-is-the-behavior-difference-between-return-path-reply-to-and-from
+	#  Email::Simple: https://metacpan.org/pod/Email::Sender::Manual::QuickStart#envelope-information
+	#
```
[the remark this answers is not in the document]
Comment (mgage) [reply lib/WeBWorK/ContentGenerator/Feedback.pm]: ok

File: lib/WeBWorK/ContentGenerator/Feedback.pm (hunk 1 above)
Comment (mgage) [reply lib/WeBWorK/ContentGenerator/Feedback.pm:313]: I'd already caught that.

File: conf/site.conf.dist [hunk 3]
```diff
@@ -287,6 +287,8 @@ $mail{smtpServer} = '';  # e.g. 'mail.yourschool.edu' or 'localhost'
 # message. It can really be anything, but some mail servers require it contain
 # a valid mail domain, or at least be well-formed.
 $mail{smtpSender} = '';  # e.g.  'webwork@yourserver.yourschool.edu'
+$mail{set_return_path}="noreply\@your.school.edu"; # or to read bounce errors set to  $mail{smptSender}
```
[the remark this answers is not in the document]
Comment (mgage) [reply conf/site.conf.dist]: ok for text entry.  You'll need to use \@  and double quotes for "noreply\@$mail{smptServer}"

[the remark this answers is not in the document]
Comment (drgrice1) [reply conf/site.conf.dist]: That is true.  You will also need to escape the @ in that case.

**Comment by mgage** [comment]
The modifications at 
after line 893 SendMail.pm
after line 949 SendMail.pm
after line 619 ProblemUtil.pm
have been made.
Some clean up of comments and indenting have been done.

File: conf/site.conf.dist [hunk 4]
```diff
@@ -287,8 +287,14 @@ $mail{smtpServer} = '';  # e.g. 'mail.yourschool.edu' or 'localhost'
 # message. It can really be anything, but some mail servers require it contain
 # a valid mail domain, or at least be well-formed.
 $mail{smtpSender} = '';  # e.g.  'webwork@yourserver.yourschool.edu'
-
 # Be sure to use single quotes for the address or the @ sign will be interpreted as an array.
+#
+$mail{set_return_path}='noreply@math.rochester.edu'; #discards error messages
```
Comment (drgrice1) [inline conf/site.conf.dist:292]: I think that by default this should probably be commented out.  Either that or it should be set to the empty string like the `$mail{smtpSender}` variable.  Your code will handle that correctly either way.

**Comment by pstaabp** [comment]
Question about testing:  on a dev machine, how do you set an smtp server?  I was trying to use google, but got errors probably due to lack of authentication.

**Review by pstaabp — CHANGES_REQUESTED** [review]

**Comment by mgage** [comment]
I'd also like to know how to set up an smtp server locally. -- I was using UR smtp server.

**Comment by dlglin** [comment]
It looks like postfix is the recommended SMTP server on a linux box.

**Comment by drdrew42** [comment]
I've used postfix, but only as a relay to the campus SMTP server... I was having issues with the campus server timing out and running my own postfix allowed me to queue emails instead of having them dropped by the remote SMTP and being lost without any notification.

**Comment by pstaabp** [comment]
There is a docker image on docker hub for postfix: https://hub.docker.com/r/mailu/postfix

**Review by drgrice1 — APPROVED** [review]
This looks good now.

**Pull request merged**